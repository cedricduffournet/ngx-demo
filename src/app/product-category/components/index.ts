export * from './product-category-add/product-category-add.component';
export * from './product-category-update/product-category-update.component';
export * from './product-category-delete/product-category-delete.component';
export * from './product-category-form/product-category-form.component';
export * from './product-category-item/product-category-item.component';
export * from './product-category-items/product-category-items.component';
