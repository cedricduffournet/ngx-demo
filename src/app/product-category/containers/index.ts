export * from './product-category-add-modal/product-category-add-modal.component';
export * from './product-category-update-modal/product-category-update-modal.component';
export * from './product-category-delete-modal/product-category-delete-modal.component';
export * from './product-category-list-view/product-category-list-view.component';
