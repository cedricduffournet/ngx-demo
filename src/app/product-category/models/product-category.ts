import { schema } from 'normalizr';

export interface ProductCategory {
  id: number;
  name: string;
}

export const productCategorySchema = new schema.Entity('productCategories');
