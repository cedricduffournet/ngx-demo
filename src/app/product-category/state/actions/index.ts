import * as ProductCategoryUpdateModalActions from './product-category-update-modal.actions';
import * as ProductCategoryAddModalActions from './product-category-add-modal.actions';
import * as ProductCategoryDeleteModalActions from './product-category-delete-modal.actions';
import * as ProductCategoryListViewActions from './product-category-list-view.actions';
import * as ProductCategoryApiActions from './product-category-api.actions';

export {
  ProductCategoryListViewActions,
  ProductCategoryApiActions,
  ProductCategoryUpdateModalActions,
  ProductCategoryAddModalActions,
  ProductCategoryDeleteModalActions
};
