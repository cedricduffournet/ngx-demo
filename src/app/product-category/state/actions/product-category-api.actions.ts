import { props, createAction } from '@ngrx/store';

import { NormalizedData } from '@app/shared/models/normalized.model';

export const loadProductCategorySuccess = createAction(
  '[ProductCategory API] Load SUCCESS',
  props<{ productCategories: NormalizedData }>()
);

export const loadProductCategoryFailure = createAction(
  '[ProductCategory API] Load FAILURE',
  props<{ error: any }>()
);

export const addProductCategorySuccess = createAction(
  '[ProductCategory API] Add SUCCESS',
  props<{ productCategory: NormalizedData }>()
);

export const addProductCategoryFailure = createAction(
  '[ProductCategory API] Add FAILURE',
  props<{ error: any }>()
);

export const updateProductCategorySuccess = createAction(
  '[ProductCategory API] Update  SUCCESS',
  props<{ productCategory: NormalizedData }>()
);

export const updateProductCategoryFailure = createAction(
  '[ProductCategory API] Update FAILURE',
  props<{ error: any }>()
);

export const deleteProductCategorySuccess = createAction(
  '[ProductCategory API] Delete SUCCESS',
  props<{ id: number }>()
);

export const deleteProductCategoryFailure = createAction(
  '[ProductCategory API] Delete FAILURE',
  props<{ error: any }>()
);
