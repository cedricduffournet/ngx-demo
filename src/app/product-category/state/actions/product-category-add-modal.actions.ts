import { props, createAction } from '@ngrx/store';

import { ProductCategory } from '@app/product-category/models/product-category';

export const addProductCategory = createAction(
  '[ProductCategory add  modal] Add',
  props<{ productCategory: ProductCategory }>()
);
