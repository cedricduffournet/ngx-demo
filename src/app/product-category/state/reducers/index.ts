import {
  createSelector,
  createFeatureSelector,
  Action,
  combineReducers
} from '@ngrx/store';

import * as fromRoot from '@app/core/state/reducers';
import * as fromProductCategoryCollection from '@app/product-category/state/reducers/product-category-collection.reducer';
import * as fromProductCategoryEntities from '@app/product-category/state/reducers/product-category-entities.reducer';
import * as fromAuth from '@app/authentication/state/reducers';
import { ProductCategory } from '@app/product-category/models/product-category';
import { Authorization } from '@app/core/models/authorization.model';

export interface ProductCategoriesState {
  collection: fromProductCategoryCollection.State;
  productCategories: fromProductCategoryEntities.State;
}

export interface State extends fromRoot.State {
  productCategories: ProductCategoriesState;
}

export function reducers(state: ProductCategoriesState | undefined, action: Action) {
  return combineReducers({
    collection: fromProductCategoryCollection.reducer,
    productCategories: fromProductCategoryEntities.reducer
  })(state, action);
}

export const selectProductCategoriesState = createFeatureSelector<
  State,
  ProductCategoriesState
>('productCategories');

export const getProductCategoryEntitiesState = createSelector(
  selectProductCategoriesState,
  state => state.productCategories
);

export const getProductCategoryEntities = createSelector(
  getProductCategoryEntitiesState,
  fromProductCategoryEntities.getEntities
);

export const getProductCategoryEntitiesUpdating = createSelector(
  getProductCategoryEntitiesState,
  fromProductCategoryEntities.getUpdating
);

export const getProductCategoryEntitiesUpdated = createSelector(
  getProductCategoryEntitiesState,
  fromProductCategoryEntities.getUpdated
);

export const getProductCategoryCollectionState = createSelector(
  selectProductCategoriesState,
  state => state.collection
);
export const getProductCategoryIds = createSelector(
  getProductCategoryCollectionState,
  fromProductCategoryCollection.getIds
);
export const getProductCategoryCollectionAdding = createSelector(
  getProductCategoryCollectionState,
  fromProductCategoryCollection.getAdding
);
export const getProductCategoryCollectionAdded = createSelector(
  getProductCategoryCollectionState,
  fromProductCategoryCollection.getAdded
);
export const getProductCategoryCollectionDeleting = createSelector(
  getProductCategoryCollectionState,
  fromProductCategoryCollection.getDeleting
);
export const getProductCategoryCollectionDeleted = createSelector(
  getProductCategoryCollectionState,
  fromProductCategoryCollection.getDeleted
);

export const getProductCategories = createSelector(
  getProductCategoryEntities,
  getProductCategoryIds,
  (entities, ids) => {
    return ids.map(id => entities[id]);
  }
);

export const getSelectedProductCategoryId = createSelector(
  getProductCategoryEntitiesState,
  fromProductCategoryEntities.getSelectedId
);

export const getSelectedProductCategory = createSelector(
  getProductCategoryEntities,
  getSelectedProductCategoryId,
  (productCategoryEntities, productCategoryId): ProductCategory => {
    return productCategoryEntities[productCategoryId];
  }
);

export const canUpdateProductCategory = createSelector(
  fromAuth.getAuthorized(['ROLE_PRODUCT_CATEGORY_EDIT']),
  canUpdate => canUpdate
);

export const canDeleteProductCategory = createSelector(
  fromAuth.getAuthorized(['ROLE_PRODUCT_CATEGORY_DELETE']),
  canDelete => canDelete
);

export const canCreateProductCategory = createSelector(
  fromAuth.getAuthorized(['ROLE_PRODUCT_CATEGORY_CREATE']),
  canCreate => canCreate
);

export const getProductCategoryAuthorization = createSelector(
  canUpdateProductCategory,
  canDeleteProductCategory,
  canCreateProductCategory,
  (canUpdate, canDelete, canAdd): Authorization => {
    return {
      update: canUpdate,
      delete: canDelete,
      create: canAdd
    };
  }
);
