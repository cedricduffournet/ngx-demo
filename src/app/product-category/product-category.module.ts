import { NgModule, ComponentFactoryResolver } from '@angular/core';

import { RouterModule } from '@angular/router';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';

import { SharedModule } from '@app/shared/shared.module';
import { ProductCategoryEffects } from '@app/product-category/state/effects';
import {
  ProductCategoryAddComponent,
  ProductCategoryUpdateComponent,
  ProductCategoryDeleteComponent,
  ProductCategoryFormComponent,
  ProductCategoryItemComponent,
  ProductCategoryItemsComponent
} from '@app/product-category/components';
import {
  ProductCategoryAddModalComponent,
  ProductCategoryUpdateModalComponent,
  ProductCategoryDeleteModalComponent,
  ProductCategoryListViewComponent
} from '@app/product-category/containers';
import { reducers } from '@app/product-category/state/reducers';
import { CoalescingComponentFactoryResolver } from '@app/coalescing-component-factory-resolver.service';

@NgModule({
  imports: [
    RouterModule,
    SharedModule,
    EffectsModule.forFeature([ProductCategoryEffects]),
    StoreModule.forFeature('productCategories', reducers)
  ],
  declarations: [
    ProductCategoryListViewComponent,
    ProductCategoryAddComponent,
    ProductCategoryUpdateComponent,
    ProductCategoryDeleteComponent,
    ProductCategoryFormComponent,
    ProductCategoryItemComponent,
    ProductCategoryItemsComponent,
    ProductCategoryAddModalComponent,
    ProductCategoryUpdateModalComponent,
    ProductCategoryDeleteModalComponent
  ],
  entryComponents: [
    ProductCategoryUpdateModalComponent,
    ProductCategoryAddModalComponent,
    ProductCategoryDeleteModalComponent
  ],
  exports: [ProductCategoryListViewComponent]
})
export class ProductCategoryModule {
  // see https://github.com/angular/angular/issues/14324
  constructor(
    coalescingResolver: CoalescingComponentFactoryResolver,
    localResolver: ComponentFactoryResolver
  ) {
    coalescingResolver.registerResolver(localResolver);
  }
}
