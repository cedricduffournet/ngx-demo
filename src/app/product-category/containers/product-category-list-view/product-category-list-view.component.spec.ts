import { TestBed, ComponentFixture } from '@angular/core/testing';

import { Store } from '@ngrx/store';
import { MockStore, provideMockStore } from '@ngrx/store/testing';
import { TranslateModule } from '@ngx-translate/core';

import {
  ProductCategoryItemComponent,
  ProductCategoryItemsComponent
} from '@app/product-category/components';
import { ProductCategoryListViewComponent } from '@app/product-category/containers';
import * as fromProductCategories from '@app/product-category/state/reducers';
import { ProductCategoryListViewActions } from '@app/product-category/state/actions';
import { SharedModule } from '@app/shared/shared.module';

describe('ProductCategoryListViewComponent', () => {
  let fixture: ComponentFixture<ProductCategoryListViewComponent>;
  let component: ProductCategoryListViewComponent;
  let store: MockStore<fromProductCategories.State>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [
        ProductCategoryListViewComponent,
        ProductCategoryItemsComponent,
        ProductCategoryItemComponent
      ],
      imports: [SharedModule, TranslateModule.forRoot()],
      providers: [
        provideMockStore({
          selectors: [
            {
              selector: fromProductCategories.getProductCategories,
              value: []
            },
            {
              selector: fromProductCategories.getProductCategoryAuthorization,
              value: {
                create: true,
                delete: true,
                update: true
              }
            }
          ]
        })
      ]
    });

    fixture = TestBed.createComponent(ProductCategoryListViewComponent);
    component = fixture.componentInstance;
    store = TestBed.get(Store);
    spyOn(store, 'dispatch');
  });

  it('should be created', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch loadProductCategories on init', () => {
    const action = ProductCategoryListViewActions.loadProductCategories();
    fixture.detectChanges();
    expect(store.dispatch).toHaveBeenCalledWith(action);
  });

  it('should dispatch showAddProductCategoryModal on add event', () => {
    const action = ProductCategoryListViewActions.showAddProductCategoryModal();
    component.onAdd();
    expect(store.dispatch).toHaveBeenCalledWith(action);
  });

  it('should dispatch showUpdateProductCategoryModal on update event', () => {
    const action = ProductCategoryListViewActions.showUpdateProductCategoryModal();
    component.onUpdate(1);
    expect(store.dispatch).toHaveBeenCalledWith(action);
  });

  it('should dispatch selectProductCategory on update event', () => {
    const action = ProductCategoryListViewActions.selectProductCategory({ id: 1 });
    component.onUpdate(1);
    expect(store.dispatch).toHaveBeenCalledWith(action);
  });

  it('should dispatch showDeleteProductCategoryModal on update event', () => {
    const action = ProductCategoryListViewActions.showDeleteProductCategoryModal();
    component.onDelete(1);
    expect(store.dispatch).toHaveBeenCalledWith(action);
  });

  it('should dispatch selectProductCategory on deletee event', () => {
    const action = ProductCategoryListViewActions.selectProductCategory({ id: 1 });
    component.onDelete(1);
    expect(store.dispatch).toHaveBeenCalledWith(action);
  });
});
