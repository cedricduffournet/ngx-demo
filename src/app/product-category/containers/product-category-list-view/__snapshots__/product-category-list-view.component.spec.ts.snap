// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCategoryListViewComponent should be created 1`] = `
<app-product-category-list-view
  authorization$={[Function Store]}
  loggedUser$={[Function Store]}
  productCategories$={[Function Store]}
  store={[Function MockStore]}
>
  <app-page-inner>
    <div
      class="px-3 py-3 "
    >
      <div
        class="page-section"
      >
        <app-product-category-items
          ng-reflect-authorization="[object Object]"
          ng-reflect-product-categories=""
        >
          <section
            class="bg-white border-left border-right border-bottom rounded"
          >
            <div
              class="table-responsive"
            >
              <table
                class="table table-hover"
              >
                <thead>
                  <tr>
                    <th
                      class="align-middle"
                      ng-reflect-translate=""
                      translate=""
                    >
                       NAME 
                    </th>
                    <th
                      class="align-middle text-right"
                    >
                       Actions 
                    </th>
                  </tr>
                </thead>
                <tbody>
                  
                </tbody>
              </table>
            </div>
          </section>
          
          <app-button-circle-right
            label="+"
            ng-reflect-label="+"
          >
            <div>
              <button
                class="btn-circle btn-xl btn btn-absolute-corner-right btn-success"
                ng-reflect-klass="btn-circle btn-xl btn btn-abso"
                ng-reflect-ng-class="btn-success"
                type="button"
              >
                 + 
              </button>
            </div>
          </app-button-circle-right>
        </app-product-category-items>
      </div>
    </div>
  </app-page-inner>
</app-product-category-list-view>
`;
